### ISSUE TITLE
GPU and CPU utilization slashed after first epoch using tf.data and tf.keras

### ISSUE THREAD

**Issue by trzy** [tensorflow/tensorflow#45114]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): v1.12.1-46274-g410a6d8f46 2.5.0-dev20201121
- Python version: 3.8.6
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: CUDA 11.1/cuDNN-11.1-8.0.5.39
- GPU model and memory: Nvidia RTX 3090 24GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

Training performance of a 16-layer CNN (VGG16) decreases dramatically after the first epoch. I am using tf.data to load ImageNet. The first epoch performs as well as can be expected on my machine (training time ~1hr), with CPU utilization of 65% and GPU at 88%. 

Epoch 2 has CPU utilization of 40.5% with the GPU at ~35% (usage appears spiky, indicating some sort of bottle neck that is idling the GPU). Data is being loaded from an NVMe SSD and there is no discernable reason for performance to be reduced on the second epoch (i.e., SSD temperature is not abnormally high). I suspect tf.data is the culprit but I don't know why -- this code works on a different system (read on below).

The dataset is ImageNet. 

**Describe the expected behavior**

There should be no performance difference between Epoch 1 and subsequent epochs (subject to thermal throttling, which I have confirmed is not an issue on my system). When running this same code on a different system (w/ Titan V GPU and HDD), performance is sustained. The version of TF there is different: v2.2.0-rc4-8-g2b96f3662b 2.2.0

The official TF2 package does not appear to support RTX 3090 GPUs on Windows, which require CUDA 11.1. Hence, I used a tf-nightly-gpu release.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

Complete code is attached.

The relevant bit is probably the tf.data dataset construction and how the fit command is executed:

```
# Construct tf.data.Dataset
paths_ds = tf.data.Dataset.from_tensor_slices(image_paths)
paths_ds = paths_ds.shuffle(buffer_size=len(image_paths), reshuffle_each_iteration=True)  # reshuffle each epoch
dataset = paths_ds.map(
  lambda path: (self.load_image(path), self.get_onehot_label(path, tf_class_index_by_image, self.num_classes)),
  num_parallel_calls=num_threads
)
dataset = dataset.batch(batch_size)
dataset = dataset.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)  #TODO: why does this sometimes slow things down?
self.dataset = dataset
```

The intent of this pipeline is to shuffle the dataset completely each epoch by shuffling the file names. Then, I would like to prefetch and load -- in parallel -- as many images as possible.

```
model.fit(callbacks=callbacks_list, x=data.dataset, epochs=options.epochs)
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

Windows performance monitoring screenshots are attached. Please note the difference between CPU_Epoch1 / GPU_Epoch1 and CPU_Epoch2 / GPU_Epoch2 image sets.

Sample code: [VGG_Bug_Sample_Code.txt](https://github.com/tensorflow/tensorflow/files/5585368/VGG_Bug_Sample_Code.txt)

CPU Epoch 1:

![CPU_Epoch1](https://user-images.githubusercontent.com/1285573/100008084-d3a88580-2d81-11eb-85bf-0f85b413ceff.jpg)

GPU Epoch 1:

![GPU_Epoch1](https://user-images.githubusercontent.com/1285573/100008115-dc00c080-2d81-11eb-8eff-0db5be3e4297.jpg)

CPU Epoch 2 (reduced utilization):

![CPU_Epoch2](https://user-images.githubusercontent.com/1285573/100008098-d73c0c80-2d81-11eb-99cc-bdf9026ca071.jpg)

GPU Epoch 2 (substantial bottleneck visible):

![GPU_Epoch2](https://user-images.githubusercontent.com/1285573/100008137-e0c57480-2d81-11eb-9d34-3eae6ba67d59.jpg)

**Comment by Saduf2019** [comment]
@trzy 
I ran the code shared on tf 2.3 and face a [different error](https://colab.research.google.com/gist/Saduf2019/35e3437143aea16a3bdaec2f843057cb/untitled471.ipynb), can you please share a colab gist with the error reported.

**Comment by trzy** [comment]
This error occurs if the dataset directory is incorrect. Do you have access to ImageNet (it needs to be preprocessed a bit) or any other image data set? If not I will try to see if I can generate synthetic data in the loading stage.

**Comment by trzy** [comment]
I tried generating synthetic data to replicate the problem but was unable to (1.2M simple JPEGs at 6GB total is not equivalent to 1.2M JPEGs occupying 160GB as with the ImageNet dataset). 

However, I got Tensorboard profiling up and captured some very interesting traces that highlight the problem.

Epoch 1 overview (most of the time is spent computing on the GPU):

![Epoch1_overview](https://user-images.githubusercontent.com/1285573/100284946-dabbc900-2f24-11eb-8b59-068cb3281ca8.jpg)

Epoch 2 overview (input bottleneck appears):

![Epoch2_overview](https://user-images.githubusercontent.com/1285573/100285006-f3c47a00-2f24-11eb-87d8-1989d3cfa9c0.jpg)

What's going on?

Trace view indicates that in Epoch 1, data loading for the subsequent batch happens across four threads during training of the current batch. Loading a batch takes <50% of the training time, so by the time the current batch is completed, it is ready to proceed to the next.

Here is Epoch 1's trace:

![Epoch1_prefetch](https://user-images.githubusercontent.com/1285573/100285181-4aca4f00-2f25-11eb-85cf-f0ca57f01e96.jpg)

Zooming in, we can see that the loading operation consists of reading, JPEG decoding, a cast, and a resize. They happen one after another.

![Epoch1_prefetch_mid_training](https://user-images.githubusercontent.com/1285573/100285237-633a6980-2f25-11eb-8a4f-72b7b17ef891.jpg)

Now compare Epoch 2:

![Epoch2_saturated](https://user-images.githubusercontent.com/1285573/100285287-7a795700-2f25-11eb-8659-75d3549ce997.jpg)

We can see the input pipeline has slowed down dramatically and it takes longer than a training step to load a batch.

<em>Why are their gaps between the different input pipeline steps in Epoch 2?</em> There is idle time between reading the file, decoding the JPEG, resizing, etc.

<em>tf_data_iterator_get_next is consuming more time.</em> Why? It is barely visible in Epoch 1's trace.

Is this something to do with the way my pipeline is constructed? 

Here is Epoch 2's input bottleneck analysis but I am unsure how to interpret it because I have only a map() call, not map_and_batch as shown here:

![Epoch2_bottleneck_analysis](https://user-images.githubusercontent.com/1285573/100285732-3fc3ee80-2f26-11eb-803f-d746be8b4a1a.jpg)

I hope this is more helpful.

**Comment by trzy** [comment]
Just popping in to add some more information. This problem appears to be very similar to mine: https://github.com/fizyr/keras-retinanet/issues/322

It appears that in their case, a driver update fixed the problem. Googling around has confirmed that Nvidia 30-series GPUs require CUDA 11.1, which Tensorflow is not built against. Hence, tf-nightly-gpu. This may be a 3090-specific issue and it would be nice to understand whether the problem could be within CUDA or Tensorflow.

**Comment by gowthamkpr** [comment]
@trzy Can you Try using the latest version of tensorflow 2.5.0 or above should fix your problem. Please let me know. Thanks!

**Comment by trzy** [comment]
@gowthamkpr I just tried it today and it seems to work much better. Ran it for 4 epochs and performance was consistent. It's not the exact same machine I had when I submitted the bug report but it is Win10 and the same 3090. I tried with tensorflow 2.9.1.

Any idea what changed?

**Comment by gowthamkpr** [comment]
@trzy Tensorflow 2.4.0 and above supports cuda 11.0 and above. Also it supports python 3.8. That might have been the cause of your problem. Not using the supportedversion of tensorflow for cuda version. You can find tensorflow and its related GPU support [here](https://www.tensorflow.org/install/source#gpu). Thanks!
